File: backend/src/main/java/com/lifescale/backend/common/model/ApiResponse.java
package com.lifescale.backend.common.model;

import com.fasterxml.jackson.annotation.JsonInclude;

/**
 * 通用 API 响应包装，与前端 ApiResponse&lt;T&gt; 类型对齐。
 */
@JsonInclude(JsonInclude.Include.ALWAYS)
public record ApiResponse<T>(
        int code,
        boolean success,
        String message,
        T data) {

    public static <T> ApiResponse<T> ok(T data) {
        return new ApiResponse<>(200, true, "ok", data);
    }

    public static <T> ApiResponse<T> ok() {
        return new ApiResponse<>(200, true, "ok", null);
    }

    public static <T> ApiResponse<T> fail(String message) {
        return new ApiResponse<>(500, false, message, null);
    }

    public static <T> ApiResponse<T> fail(int code, String message) {
        return new ApiResponse<>(code, false, message, null);
    }

    public static <T> ApiResponse<T> fail(int code, String message, T data) {
        return new ApiResponse<>(code, false, message, data);
    }
}

File: mobile/lib/core/network/dto/api_response.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'api_response.freezed.dart';
part 'api_response.g.dart';

/// 后端统一响应信封，与 `backend/.../common/model/ApiResponse.java` 对齐。
/// 后端 `@JsonInclude(ALWAYS)`：四字段恒在（即便 data=null）。
///
/// 泛型 `T` 经 `genericArgumentFactories` 由调用方提供 `fromJsonT`。
@Freezed(genericArgumentFactories: true)
class ApiResponse<T> with _$ApiResponse<T> {
  const factory ApiResponse({
    required int code,
    required bool success,
    required String message,
    T? data,
  }) = _ApiResponse<T>;

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object?) fromJsonT,
  ) => _$ApiResponseFromJson<T>(json, fromJsonT);
}
